src/lib/aria-roles.ts

```diff
@@ -12,922 +13,1686 @@ import type {
 
 export type RoleType = 'abstract' | 'widget' | 'document' | 'landmark' | 'liveregion' | 'window' | 'graphics';
 
+// note: all fields required to be monomorphic
 export interface RoleData {
-  type: RoleType[];
   /**
-   * If given, states that this role can only exist within this container
-   * @see https://www.w3.org/TR/wai-aria-1.3/#scope
+   * A list of roles which are allowed on an accessibility child (simplified as "child") of the element with this role.
+   * @see https://w3c.github.io/aria/#mustContain
    */
-  requiredParents?: ARIARole[];
-  superclasses?: ARIARole[]; // ignores abstract roles
-  subclasses?: ARIARole[]; // ignores abstract roles
+  allowedChildRoles: ARIARole[];
   /**
-   * Role that require an accessible name.
+   * The DOM descendants are presentational.
+   * @see https://w3c.github.io/aria/#childrenArePresentational
    */
+  childrenPresentational: boolean;
+  /** Default values for ARIA attributes (if any) */
+  defaultAttributeValues: Record<string, string>;
+  /** Which HTML elements inherit this role, if any (note: attributes may be necessary) */
+  elements: VirtualElement[];
+  name: ARIARole;
+  /** Role that require an accessible name. */
   nameRequired: boolean;
+  /** Role that prohibits having a name */
+  nameProhibited: boolean;
+  /**
+   * aria-* attributes that are explicitly prohibited for this role, and are considered an error if set.
+   * @see https://www.w3.org/TR/wai-aria-1.3/#prohibitedattributes
+   */
+  prohibited: ARIAAttribute[];
+  /**
+   * If given, states that this role can only exist within this container
+   * @see https://www.w3.org/TR/wai-aria-1.3/#scope
+   */
+  requiredParentRoles: ARIARole[];
   /**
    * aria-* attributes that MUST be set for this role.
    * @see https://www.w3.org/TR/wai-aria-1.3/#requiredState
    */
   required: ARIAAttribute[];
+  superclasses: (ARIARole | AbstractRole)[]; // ignores abstract roles
+  subclasses: (ARIARole | AbstractRole)[]; // ignores abstract roles
   /**
    * aria-* attributes that MAY be set for this role.
    * Note: this includes required attributes, supported attributes, and inherited attributes from superclass role types.
    * @see https://www.w3.org/TR/wai-aria-1.3/#supportedState
    * @see https://www.w3.org/TR/wai-aria-1.3/#inheritedattributes
    */
   supported: ARIAAttribute[];
-  /**
-   * aria-* attributes that are explicitly prohibited for this role, and are considered an error if set.
-   * @see https://www.w3.org/TR/wai-aria-1.3/#prohibitedattributes
-   */
-  prohibited: ARIAAttribute[];
-  /**
-   * Which HTML elements inherit this role, if any (note: attributes may be necessary)
-   */
-  elements: VirtualElement[] | undefined;
+  type: RoleType[];
 }
 
 export const widgetRoles: Record<WidgetRole, RoleData> = {
   /** An input that allows for user-triggered actions when clicked or pressed. See related link. */
   button: {
-    type: ['widget'],
+    allowedChildRoles: [],
```

Comment:
I read a [blogpost](https://romgrk.com/posts/optimizing-javascript) about the peformance gains of monomorphism and since I had to touch a lot of these anyway, decided why not?

Is it faster? I don’t know. We’ll never know. But it was an _attempt_ at performance